eng/targets/TargetFrameworks.props

```diff
@@ -55,7 +55,7 @@
         <NetRoslyn>$(NetCurrent)</NetRoslyn>
         <NetRoslynSourceBuild>$(NetCurrent);$(NetPrevious)</NetRoslynSourceBuild>
         <NetRoslynAll>$(NetCurrent);$(NetPrevious)</NetRoslynAll>
-        <NetRoslynBuildHostNetCoreVersion>$(NetCurrent)</NetRoslynBuildHostNetCoreVersion>
+        <NetRoslynBuildHostNetCoreVersion>$(NetPrevious)</NetRoslynBuildHostNetCoreVersion>
```

Comment:
deferring to @mthalman and @jaredpar who worked on this pattern.

Reply:
cc @jasonmalinowski

Reply:
> Shouldn't roslyn's source-built product (built inside the VMR) only target NetCurrent?

Every time we've tried this in the past we've ended up breaking some other part of the build (there is always an older runtime or dependency hanging around that forces us into `$(NetPrevious)`). This is why we keep both current + prev even when in full source build. If you all are confident this no longer holds we can remove the all the `$(NetPrevious)` from this `<When>` block ... but I will small stakes bet we will find some way this breaks us again :wink:

Alternatively if we want to keep building current + previous I think the right fix is to change the project files so that they agree on TFM as they depend on each other: 

| Project | TFM Set |
| --- | ---|
| Microsoft.CodeAnalysis.Workspaces.MSBuild.BuildHost.csproj | $(NetRoslynBuildHostNetCoreVersion);net472 | |
| Microsoft.CodeAnalysis.Workspaces.MSBuild.csproj | $(NetRoslynSourceBuild);net472 |

Basically make the latter `$(NetRoslynBuildHostNetCoreVersion)`

Reply:
Note that NetRoslynBuildHostNetCoreVersion is used as TargetFramework here

https://github.com/dotnet/roslyn/blob/c8fba927cf53c6826e7dcf272abc1ff835a890f8/src/Workspaces/Core/MSBuild/Microsoft.CodeAnalysis.Workspaces.MSBuild.csproj#L101

so that will need to be updated somehow if NetRoslynBuildHostNetCoreVersion is changed to contain multiple values.

Reply:
> Note that NetRoslynBuildHostNetCoreVersion is used as TargetFramework (no plural) here

I've been thinking that we should change our approach to have our named TFs end with an 's' if they can be multiple in any configuration. That would make it easier to read project files at a glance.

Reply:
> Note that NetRoslynBuildHostNetCoreVersion is used as TargetFramework (no plural) here

I'm not sure why that variable isn't being used in the other location; looking through history I didn't see an obvious place where it was...might have been an oversight when the variable was originally introduced?

> if NetRoslynBuildHostNetCoreVersion is changed to contain multiple values

I can't imagine a case where we'd need multiple ones; it'd be tricky anyways to have to bundle multiple versions of the BuildHost and pick the right one as it is, so assuming there would only be a single TFM seems reasonable.

Generally, NetCurrent should be fine for the SDK build, other than Jared's concern that that often creates a surprise. For our official builds that produce our NuGet packages that go to nuget.org, "older is better" in terms of TFM, especially since tools like the upgrade assistants need to potentially analyze projects for upgrade and may not have newer runtimes on the box. (I'd seriously set it to netcoreapp3.1 if it wasn't likely to result in
![image](https://github.com/user-attachments/assets/53a9e8b6-d7d4-4fe8-a429-0c2c9eb03100)

Reply:
> Generally, NetCurrent should be fine for the SDK build, other than Jared's concern that that often creates a surprise.

The tricky part about the SDK is that Roslyn simultaneously flows to both .NET preview and .NET stable. Concretely at this moment every change in `main` will flow to both 9.0.200 and 10.0. This is true for both our production and repo source build assets. 

When we target only NetCurrent in repo source build then only a NetCurrent compiler, libs, etc ... are flowed to repos still targeting 9.0.200. When they do their repo level source build they are quite possibly only using a `net9.0` SDK, or using `net9.0` only TFMS. That breaks down when we hand them compilers / libraries tied to `net10.0`.

I can't say for certain how that applies to full source build. But have vague memories of us trying to make that NetCurrent pure and failing. 

But it's somewhat immaterial for workspaces layer. That is a component that flows, hence it participates in repo source build and must abide by the NetCurrent + NetPrevious restriction.

Reply:
> When we target only NetCurrent in repo source build then only a NetCurrent compiler, libs, etc ... are flowed to repos still targeting 9.0.200. When they do their repo level source build they are quite possibly only using a net9.0 SDK, or using net9.0 only TFMS. That breaks down when we hand them compilers / libraries tied to net10.0.
I can't say for certain how that applies to full source build. But have vague memories of us trying to make that NetCurrent pure and failing.

I fully get the point why the repo source-build needs to target both `NetPrevious` and `NetCurrent`. In full product source-build, `NetCurrent` should be enough as it automatically adapts to the TFM that is required in the given context and release. I.e. in 9.0.200 NetCurrent=net9.0 but when you insert into 10.0.100, NetCurrent=net10.0.

One unrelated observation: it would be great if the repo source-build leg would also target the TFM that we insert into the (latest) product. Right now it depends on the repo SDK (9.0.100)  and target net8.0 and net9.0 (NetPrevious and NetCurrent) with Arcade 9. The latest product though builds with the 10.0.100 SDK and net10.0 TFM. To me it looks like we are missing validation here.

Reply:
> . In full product source-build, NetCurrent should be enough as it automatically adapts to the TFM that is required in the given context

In programming _should_ is a four letter word. ;)

Again, I feel like every time we try this we convince ourselves it's safe, find out that in practice it's not and we revert. I'm happy to take the bet again but it's more up to you all. 

>  it would be great if the repo source-build leg would also target the TFM t

I'm hopeful that once we get VMR source build working that we can eliminate repo source build entirely. It seems to be of diminishing value given how quickly we get full VMR validation post merge.

Reply:
> I'm hopeful that once we get VMR source build working that we can eliminate repo source build entirely. It seems to be of diminishing value given how quickly we get full VMR validation post merge.

We were actually considering the opposite by adding another repo validation leg to mimic the product build (VMR) when not building from source. I'm not sure whether we would prefer to find breaks post-merge during insertion from roslyn -> sdk. I admit that the current (source-build) validation leg doesn't validate all important cases which is why I asked above whether it should target the newer TFM and run with the newer Arcade SDK and .NET SDK.

cc @mmitche @MichaelSimons

Reply:
My mental model is that we have repo source build because the time between when a change merges and when we get full source build validation is very long. For a normal change the validation is probably ~3 days, and pretty regularly it can week(s). It's also hard to spot test a change because you really need an official build + darc flow to do a source build test. 

None of that is true with VMR source build. The time between merging a change and getting full source build validation is hours. That is because we flow code not binaries. So the moment we merge the flow should happen to VMR. Also as a dev if I think a change is source build risky it's trivial to validate locally. Just apply my change to VMR, run the source build leg (can even put up a PR) and I will get validation. 

At that point I would push back on the need for repo local validation of source build.

Reply:
I forwarded your message to our internal communication channel. Thanks for providing your input. Now is the best time to share opinions as we haven't started the repo level validation epic yet. We will discuss this in length (presumably in our next sync meeting) and get back to you.

Reply:
Happy to come to a sync and elaborate in detail about this.